massa-protocol-worker/src/worker_operations_impl.rs

```diff
@@ -0,0 +1,225 @@
+    signed::Signable,
+};
+use massa_network_exports::NetworkError;
+use massa_protocol_exports::{ProtocolError, ProtocolPoolEvent};
+use massa_time::TimeError;
+use std::time::Duration;
+use tokio::time::{sleep_until, Instant, Sleep};
+use tracing::{debug, warn};
+
+impl ProtocolWorker {
+    /// On receive a batch of operation ids `op_batch` from another `node_id`
+    /// Execute the following algorithm: [redirect to github](https://github.com/massalabs/massa/issues/2283#issuecomment-1040872779)
+    ///
+    ///```py
+    ///def process_op_batch(op_batch, node_id):
+    ///    ask_set = void HashSet<OperationId>
+    ///    future_set = void HashSet<OperationId>
+    ///    for op_id in op_batch:
+    ///        if not is_op_received(op_id):
+    ///            if (op_id not in asked_ops) or (node_id not in asked_ops(op_id)[1]):
+    ///                if (op_id not in asked_ops) or (asked_ops(op_id)[0] < now - op_batch_proc_period:
+    ///                    ask_set.add(op_id)
+    ///                    asked_ops(op_id)[0] = now
+    ///                    asked_ops(op_id)[1].add(node_id)
+    ///                else:
+    ///                    future_set.add(op_id)
+    ///    if op_batch_buf is not full:
+    ///        op_batch_buf.push(now+op_batch_proc_period, node_id, future_set)
+    ///    ask ask_set to node_id
+    ///```
+    pub(crate) async fn on_batch_operations_received(
+        &mut self,
+        op_batch: OperationIds,
+        node_id: NodeId,
+    ) -> Result<(), ProtocolError> {
+        let mut ask_set =
+            OperationIds::with_capacity_and_hasher(op_batch.len(), BuildMap::default());
+        let mut future_set =
+            OperationIds::with_capacity_and_hasher(op_batch.len(), BuildMap::default());
+        // exactitude isn't important, we want to have a now for that function call
+        let now = Instant::now();
+        for op_id in op_batch {
+            if self.checked_operations.contains(&op_id) {
+                continue;
+            }
+            let wish = match self.asked_operations.get_mut(&op_id) {
+                Some(wish) => {
+                    if wish.1.contains(&node_id) {
+                        continue; // already asked to the `node_id`
+                    } else {
+                        Some(wish) // already asked but at someone else
+                    }
+                }
+                None => None,
+            };
+            if let Some(wish) = wish {
+                // Ask now if latest ask instant < now - operation_batch_proc_period
+                // otherwise add in future_set
+                if wish.0
+                    < now
+                        .checked_sub(self.protocol_settings.operation_batch_proc_period.into())
+                        .ok_or(TimeError::TimeOverflowError)?
+                {
+                    debug!("re-ask operation {:?} asked for the first time {:?} millis ago.", op_id, wish.0.elapsed().as_millis());
```

Comment:
I propose to keep that as it can be useful

Reply:
okay